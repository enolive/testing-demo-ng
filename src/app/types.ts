export interface JokeResponse {
  value: string
}
