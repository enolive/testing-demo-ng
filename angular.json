{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "testing-demo-ng": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:application": {
          "strict": true
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:browser",
          "options": {
            "outputPath": "dist/testing-demo-ng",
            "index": "src/index.html",
            "main": "src/main.ts",
            "polyfills": "src/polyfills.ts",
            "tsConfig": "tsconfig.app.json",
            "assets": [
              "src/favicon.ico",
              "src/assets"
            ],
            "styles": [
              "src/styles.css"
            ],
            "scripts": []
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kb",
                  "maximumError": "1mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "2kb",
                  "maximumError": "4kb"
                }
              ],
              "fileReplacements": [
                {
                  "replace": "src/environments/environment.ts",
                  "with": "src/environments/environment.prod.ts"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "buildOptimizer": false,
              "optimization": false,
              "vendorChunk": true,
              "extractLicenses": false,
              "sourceMap": true,
              "namedChunks": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "configurations": {
            "production": {
              "browserTarget": "testing-demo-ng:build:production"
            },
            "development": {
              "browserTarget": "testing-demo-ng:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "browserTarget": "testing-demo-ng:build"
          }
        },
        "test": {
          "builder": "@angular-builders/jest:run",
          "options": {
            "tsConfig": "tsconfig.spec.json",
            "globalMocks": [
              "doctype",
              "styleTransform",
              "matchMedia"
            ]
          }
        },
        "lint": {
          "builder": "@angular-eslint/builder:lint",
          "options": {
            "lintFilePatterns": [
              "src/**/*.ts",
              "src/**/*.html"
            ]
          }
        }
      }
    }
  },
  "defaultProject": "testing-demo-ng",
  "cli": {
    "defaultCollection": "@angular-eslint/schematics"
  }
}
