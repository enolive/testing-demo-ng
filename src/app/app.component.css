.status {
  min-height: 100px;
}
