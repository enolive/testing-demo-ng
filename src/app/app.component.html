<header>
  <h1>Welcome to my app</h1>
  <figure>
    <img src="/assets/chuck-norris-logo.jpg" alt="Chuck Norris approves">
  </figure>
</header>
<main>
  <div role="status" class="status">
    <ng-container *ngIf="(jokeText$ | async) as jokeText; else loadingIndicator">
      <div *ngIf="jokeText.type === 'success'">{{jokeText.text}}</div>
      <div *ngIf="jokeText.type === 'error'" class="alert alert-danger">{{jokeText.error.message}}</div>
    </ng-container>
  </div>
  <button (click)="refreshJoke()" class="btn btn-primary">Refresh</button>
</main>

<ng-template #loadingIndicator>
  <div class="spinner-border">
    <span class="visually-hidden">Loading...</span>
  </div>
</ng-template>
