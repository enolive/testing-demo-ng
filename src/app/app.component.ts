import {Component, OnInit} from '@angular/core'
import {catchError, map, Observable, of} from 'rxjs'
import {HttpClient} from '@angular/common/http'
import {JokeResponse} from './types'

type JokeState =
  | { type: 'success', text: string }
  | { type: 'error', error: Error }

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css'],
})
export class AppComponent implements OnInit {
  constructor(private readonly http: HttpClient) {
  }

  jokeText$: Observable<JokeState> | undefined

  ngOnInit(): void {
    this.refreshJoke()
  }

  refreshJoke() {
    this.jokeText$ = this.http
      .get<JokeResponse>('https://api.chucknorris.io/jokes/random')
      .pipe(
        map(res => (<JokeState>{type: 'success', text: res.value})),
        catchError((e: Error) => of(<JokeState>{type: 'error', error: e})),
      )
  }
}
