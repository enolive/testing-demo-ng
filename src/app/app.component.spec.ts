import {AppComponent} from './app.component'
import {fireEvent, render, waitFor} from '@testing-library/angular'
import {HttpClient} from '@angular/common/http'
import {createMock, Mock} from '@testing-library/angular/jest-utils'
import {EMPTY, of, throwError} from 'rxjs'
import {JokeResponse} from './types'

describe('AppComponent', () => {
  let httpMock: Mock<HttpClient>

  const renderComponent = async () =>
    await render(AppComponent, {
      providers: [
        {provide: HttpClient, useValue: httpMock},
      ],
    })

  beforeEach(() => {
    httpMock = createMock(HttpClient)
    httpMock.get.mockReturnValue(EMPTY)
  })

  it('renders correctly', async () => {
    const {getByRole} = await renderComponent()

    expect(getByRole('heading')).toHaveTextContent('Welcome to my app')
    expect(getByRole('img')).toMatchSnapshot()
    expect(getByRole('button')).toHaveTextContent('Refresh')
    expect(getByRole('status')).toHaveTextContent('Loading...')
  })

  it('fetches random joke on load', async () => {
    const response: JokeResponse = {
      value: 'ðŸ‘Š BAM ðŸ‘Š',
    }
    httpMock.get.mockReturnValue(of(response))

    const {getByRole} = await renderComponent()

    await waitFor(() =>
      expect(getByRole('status')).toHaveTextContent('ðŸ‘Š BAM ðŸ‘Š'),
    )
    expect(httpMock.get).toHaveBeenCalledWith('https://api.chucknorris.io/jokes/random')
  })

  it('fetches random joke on button click', async () => {
    const response1: JokeResponse = {value: 'ðŸ‘Š BAM ðŸ‘Š'}
    const response2: JokeResponse = {value: 'ðŸ‘Š BOOM ðŸ‘Š'}
    httpMock.get
      .mockReturnValueOnce(of(response1))
      .mockReturnValueOnce(of(response2))
    const {getByRole} = await renderComponent()
    fireEvent.click(getByRole('button'))

    await waitFor(() =>
      expect(getByRole('status')).toHaveTextContent('ðŸ‘Š BOOM ðŸ‘Š'),
    )
  })

  it('displays errors', async () => {
    const response = throwError(() => new Error('NOOOO!'))
    httpMock.get.mockReturnValue(response)

    const {getByRole} = await renderComponent()

    await waitFor(() =>
      expect(getByRole('status')).toHaveTextContent('NOOOO!'),
    )
  })
})
