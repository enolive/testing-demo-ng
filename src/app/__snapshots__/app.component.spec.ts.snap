// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent renders correctly 1`] = `
<img
  alt="Chuck Norris approves"
  src="/assets/chuck-norris-logo.jpg"
/>
`;
